## &#128640; Team4_유니콘기업분석

### &#9989; EDA 요약

- 연매출액, 월별 연금보험료, 월별 직원수가 큰 회사, 즉 회사의 규모가 커지면 커질수록 상위 그룹의 회사에서는 값의 분포가 크게 나타난다.
- 연매출액이 가장 큰 규모의 회사들 중 1인당 매출액이 적을수록 월별 연금 보험료와 월별 직원수가 더 많아진다.
- 직원이 많을수록 급여(고정비) 대비 매출 가치 낮은 것으로 추측된다.
------
#### 1. 따라서 적정 수준의 직원과 연금보험료를 내면서도 높은 매출액을 유지하는 회사가 가치가 큰 것으로 보인다.<br>2. 연매출액 기준 규모가 큰 기업들을 선분류하고, 그 안에서 미시적인 분석을 진행한다.<br>3. 회사의 규모가 큰 그룹일수록 데이터의 분포가 매우 넓기 때문에, 성장률에 초점을 맞춰 분석을 진행한다.
------
### &#10010; EDA 상세

In [196]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()

# (Windows OS) 한글 출력을 위한 글꼴 설정
plt.rcParams['font.family'] = 'Malgun Gothic'  

# (MAC OS)
#plt.rcParams['font.family'] = 'AppleGothic'  

# 문자 - (마이너스) 정상 출력을 위한 코드
plt.rcParams['axes.unicode_minus'] = False

# 경고 제거
plt.rcParams.update({'figure.max_open_warning': 0})

In [197]:
# 데이터 파일 read (각자의 경로 입력)
wanted_nps_df=pd.read_csv('C:/Users/kang/onboarding/company_nps_data.csv', encoding='cp949')

In [198]:
wanted_nps_df.info()
wanted_nps_df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99801 entries, 0 to 99800
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype
---  ------    --------------  -----
 0   회사ID      99801 non-null  int64
 1   연매출액      99801 non-null  int64
 2   년도        99801 non-null  int64
 3   월         99801 non-null  int64
 4   월별_연금보험료  99801 non-null  int64
 5   월별_직원수    99801 non-null  int64
dtypes: int64(6)
memory usage: 4.6 MB


회사ID          연매출액            년도             월      월별_연금보험료  \
count  9.980100e+04  9.980100e+04  99801.000000  99801.000000  9.980100e+04   
mean   2.762982e+05  3.053589e+07   2016.967125      6.714352  9.734079e+06   
std    1.339032e+05  3.016647e+08      0.976688      3.512966  4.347256e+07   
min    6.300000e+01 -7.882176e+06   2015.000000      1.000000  0.000000e+00   
25%    1.655300e+05  1.409760e+06   2016.000000      4.000000  1.181160e+06   
50%    2.591490e+05  3.573585e+06   2017.000000      7.000000  2.332760e+06   
75%    3.826560e+05  9.582321e+06   2018.000000     10.000000  5.591860e+06   
max    5.159723e+06  1.132878e+10   2019.000000     12.000000  9.780673e+08   

             월별_직원수  
count  99801.000000  
mean      37.623431  
std      146.947194  
min        0.000000  
25%        6.000000  
50%       11.000000  
75%       24.000000  
max     3133.000000

### 데이터 정리
* 직원이 0명인 경우와 연매출액이 마이너스인 경우 제외 처리
* 가져온 데이터를 추가 활용하기 위해 직원 1명당 매출액을 보여주는 컬럼 추가

In [199]:
# 이상치 제외
wanted_nps_df = wanted_nps_df.drop(index=wanted_nps_df.query('월별_직원수 == 0').index)

In [200]:
wanted_nps_df

회사ID      연매출액    년도   월  월별_연금보험료  월별_직원수
0      233757  41688077  2018   7   8912380      36
1      123464   1760695  2018   3    942380       5
2      342124   3221341  2018   7    781180       4
3      386420   4815584  2015  11   3795900      14
4      129243   9799736  2018  10  40724680     151
...       ...       ...   ...  ..       ...     ...
99796  171439   4630622  2017   2   2402740      11
99797  239388   1045021  2018  10   1428800       9
99798  272304    181898  2018   3    392560       4
99799  362194   2100218  2018  12    874780       3
99800  124281    132900  2017  11    291940       5

[99788 rows x 6 columns]

In [23]:
# 직원 1인당 매출액 컬럼 추가
#wanted_nps_df.loc[:,'인당_매출액'] = [ revenue // num for revenue, num in zip(wanted_nps_df.loc[:, '연매출액'], wanted_nps_df.loc[:, '월별_직원수']) ]

# **연도별로 추출하기 위해 실행 하지 않음**

In [201]:
# 데이터의 통계량
wanted_nps_df.describe()

회사ID          연매출액            년도             월      월별_연금보험료  \
count  9.978800e+04  9.978800e+04  99788.000000  99788.000000  9.978800e+04   
mean   2.762820e+05  3.053943e+07   2016.967140      6.714515  9.735347e+06   
std    1.338940e+05  3.016841e+08      0.976691      3.512956  4.347525e+07   
min    6.300000e+01 -7.882176e+06   2015.000000      1.000000  9.000000e+04   
25%    1.655300e+05  1.409760e+06   2016.000000      4.000000  1.181860e+06   
50%    2.591490e+05  3.573585e+06   2017.000000      7.000000  2.333680e+06   
75%    3.826560e+05  9.584106e+06   2018.000000     10.000000  5.591960e+06   
max    5.159723e+06  1.132878e+10   2019.000000     12.000000  9.780673e+08   

             월별_직원수  
count  99788.000000  
mean      37.628332  
std      146.956138  
min        3.000000  
25%        6.000000  
50%       11.000000  
75%       24.000000  
max     3133.000000

In [202]:
# 회사ID 고유 건수
len(set(wanted_nps_df['회사ID']))

4998

### 데이터 통계
* 4998개 기업의 2015년~2019년까지의 데이터 (날짜는 기업별 상이)
* (참고) 각 Row 기준의 통계 / 회사, 일자 구분 X
***
- 최대 연매출액: 약 11조 3천억원
- 최대 월별 연금보험료: 약 9억 7천만원
- 최대 월별 직원수: 약 3100명
***
- 평균 연매출액: 약 300억원
- 평균 월별 연금보험료: 약 970만원
- 평균 월별 직원수: 약 37명
***
- 최소 연매출액: 약 30만원
- 최소 월별 연금보험료: 약 9만원
- 최소 월별 직원수: 3명

### 강성규 PART 

In [203]:
# 연도별 회사 조회 함수

def sorted_year(df, year):
    df = df.query(f'년도 == {year}')
    df = df.sort_values(by=['회사ID', '년도', '월'])
    df = df.reset_index(drop=True)
    return df

df_2015 = sorted_year(wanted_nps_df, 2015)
df_2015

회사ID     연매출액    년도   월  월별_연금보험료  월별_직원수
0      12767  4508357  2015  11   3284580      17
1      12767  4508357  2015  12   3284580      17
2      14271  1199495  2015  11   2571840      11
3      14271  1199495  2015  12   3558780      15
4      59533   194704  2015  11    450440       5
...      ...      ...   ...  ..       ...     ...
3581  515562   254818  2015  12    160360       3
3582  515632   208343  2015  11    442700       4
3583  515632   208343  2015  12    442700       4
3584  515837   770107  2015  11    624120       5
3585  515837   770107  2015  12    527920       4

[3586 rows x 6 columns]

In [204]:
# 평균 직원 수 column 추가

# '월', '월별_연금보험료'를 제외한 새로운 DataFrame 생성
employ_avg_df_2015 =  pd.DataFrame(df_2015[['회사ID', '연매출액', '년도']], columns= ['회사ID', '연매출액', '년도', '연평균_직원수'])
employ_avg_df_2015 = employ_avg_df_2015.drop_duplicates(keep = 'first')


# 방법 1: for loop 
avg = []
for id in df_2015['회사ID'].unique():
    result = df_2015.query(f'회사ID == {id}').loc[:, '월별_직원수'].mean()
    avg.append(int(result)) # 사람 수 이기 때문에 소수점 아래는 버림

employ_avg_df_2015['연평균_직원수'] = avg


# 방법 2: list comprehention
# employ_avg_df_2015['연평균_직원수'] = [ int(df_2015.query(f'회사ID == {id}').loc[:, '월별_직원수'].mean()) for id in df_2015['회사ID'].unique() ]


# 연매출액을 연평균_직원 수로 나누고, 내림차순으로 정렬합니다.
employ_avg_df_2015['1인당_매출액(천원)'] = employ_avg_df_2015['연매출액'] // employ_avg_df_2015['연평균_직원수'] # 소수점 버림
employ_avg_df_2015 = employ_avg_df_2015.sort_values(by=['1인당_매출액(천원)'], ascending=False) # ascending=False 내림차순

employ_avg_df_2015.head()

회사ID        연매출액    년도  연평균_직원수  1인당_매출액(천원)
2159  297874  1013847581  2015       31     32704760
3082  433492   113314185  2015        4     28328546
610   146654   661861166  2015       29     22822798
3062  430507    66498785  2015        3     22166261
3084  433756   115171896  2015        6     19195316

In [205]:
# "연평균 직원 수 column 추가" 함수

def employ_avg_extract(df):
    avg_df = pd.DataFrame(df[['회사ID', '연매출액', '년도']], columns= ['회사ID', '연매출액', '년도', '연평균_직원수'])
    avg_df = avg_df.drop_duplicates(keep = 'first')
    
    avg = []
    # 방법 1: for loop 
    for id in avg_df['회사ID'].unique():
        result = df.query(f'회사ID == {id}').loc[:, '월별_직원수'].mean()
        avg.append(int(result)) # 사람 수 이기 때문에 소수점 아래는 버림
    
    avg_df['연평균_직원수'] = avg
    avg_df['1인당_매출액(천원)'] = avg_df['연매출액'] // avg_df['연평균_직원수'] # 소수점 버림
    avg_df = avg_df.sort_values(by=['1인당_매출액(천원)'], ascending=False) # ascending=False 내림차순
    avg_df = avg_df.reset_index(drop=True)
    return avg_df

In [206]:
#( 예시 ) 함수를 이용한 데이터 조회 

df_2015 = sorted_year(wanted_nps_df, 2015)
employ_avg_df = employ_avg_extract(df_2015)
employ_avg_df.head()

회사ID        연매출액    년도  연평균_직원수  1인당_매출액(천원)
0  297874  1013847581  2015       31     32704760
1  433492   113314185  2015        4     28328546
2  146654   661861166  2015       29     22822798
3  430507    66498785  2015        3     22166261
4  433756   115171896  2015        6     19195316

In [207]:
# 연도 별 1인당_매출액 상위 5개 기업을 모은 DataFrame 생성
# 조회하는 데 시간이 다소 소요됩니다.

all_employ_avg_df = pd.DataFrame()
year_list = sorted(wanted_nps_df['년도'].unique())

for year in year_list:
    df_year = sorted_year(wanted_nps_df, year)
    employ_avg_df = employ_avg_extract(df_year)
    all_employ_avg_df = all_employ_avg_df.append(employ_avg_df.head(), ignore_index = True)

all_employ_avg_df.head(10)

회사ID        연매출액    년도  연평균_직원수  1인당_매출액(천원)
0  297874  1013847581  2015       31     32704760
1  433492   113314185  2015        4     28328546
2  146654   661861166  2015       29     22822798
3  430507    66498785  2015        3     22166261
4  433756   115171896  2015        6     19195316
5  433492   152354246  2016        3     50784748
6  297874   894075653  2016       32     27939864
7  439585    77263316  2016        3     25754438
8  146654   616259550  2016       27     22824427
9  432911    87545408  2016        6     14590901

In [208]:
# 연도별 1인당 매출액 상위 5개 기업, 등장 횟수
all_employ_avg_df.회사ID.value_counts().head(10)

146654    4
297874    4
433492    2
439585    2
102971    1
305146    1
294649    1
343863    1
40860     1
433756    1
Name: 회사ID, dtype: int64

In [209]:
# Q. 연평균_직원수의 평균으로 총 직원 수의 평균을 도출하는 게 좋을지?
for id in (146654, 297874, 433492, 439585):
    print(id, all_employ_avg_df.query(f'회사ID == {id}')['연평균_직원수'].mean())

146654 26.75
297874 36.0
433492 3.5
439585 3.0


In [210]:
# Q. 월별_직원수의 평균으로 총 직원 수의 평균을 도출하는 게 좋을지?
for id in (146654, 297874, 433492, 439585):
    print(id, wanted_nps_df.query(f'회사ID == {id}')['월별_직원수'].mean())

146654 26.55263157894737
297874 37.5
433492 4.076923076923077
439585 3.0


## 1. 연도별 1인당 매출액 상위 5개 기업, 등장 횟수
- 4년 Top 5: 146654(15-3, 16-4, 17-3, 18-3), 297874(15-1, 16-2, 17-1, 18-2) *연도-순위
- 3년 Top 5: None
- 2년 Top 5: 433492(15-2, 16-1 년), 439585(16-3, 17-2 년)
- 1년 Top 5: 나머지 전부
---
## 2. 4, 2년 상위 기업 평균 직원 수( 월별_직원수 기준 )
- 146654(4년): 26명 (26.55263157894737)
- 297874(4년): 37명 (37.5)
- 433492(2년): 4명 (4.076923076923077)
- 439585(2년): 3 명 (3.0)

In [180]:
# 1인당 매출 상위 기업
all_employ_avg_df.sort_values(by=['년도', '1인당_매출액(천원)'], ascending=False).head() # 전부 다 보고 싶으면 head 제거

회사ID         연매출액    년도  연평균_직원수  1인당_매출액(천원)
20  155754     43590434  2019       11      3962766
21  419945  10058868796  2019     2781      3616997
22  343863      9892108  2019        4      2473027
23  212335      7111215  2019        3      2370405
24  305146     32791140  2019       16      2049446

In [193]:
# query에 연도를 삽입하여 데이터 조회
sample = all_employ_avg_df.sort_values(by=['년도', '1인당_매출액(천원)'], ascending=False).query('년도 ==2019')
sample = sample.reset_index(drop=True)
sample

회사ID         연매출액    년도  연평균_직원수  1인당_매출액(천원)
0  155754     43590434  2019       11      3962766
1  419945  10058868796  2019     2781      3616997
2  343863      9892108  2019        4      2473027
3  212335      7111215  2019        3      2370405
4  305146     32791140  2019       16      2049446

---
### 2015년 1인당 매출 상위 기업( 3586 개 기업 중 )
|index|회사ID | 연매출액| 년도 | 연평균_직원수 | 1인당_매출액(천원) |
|--|-------|-------|-------|-------|-------|
|0|297874|1013847581|2015|31|32704760|
|1|433492|113314185|2015|4|28328546|
|2|146654|661861166|2015|29|22822798|
|3|430507|66498785|2015|3|22166261|
|4|433756|115171896|2015|6|19195316|

### 2016년 1인당 매출 상위 기업 ( 37678 개 기업 중 )
|index|회사ID | 연매출액| 년도 | 연평균_직원수 | 1인당_매출액(천원) |
|--|-------|-------|-------|-------|-------|
|0|	433492|	152354246|	2016|	3|	50784748|
|1|	297874|	894075653|	2016|	32|	27939864|
|2|	439585|	77263316|	2016|	3|	25754438|
|3|	146654|	616259550|	2016|	27|	22824427|
|4|	432911|	87545408|	2016|	6|	14590901|

### 2017년 1인당 매출 상위 기업( 17805 개 기업 중 )
|index|회사ID | 연매출액| 년도 | 연평균_직원수 | 1인당_매출액(천원) |
|--|-------|-------|-------|-------|-------|
|0|	297874|	1501348760|	2017|	40|	37533719|
|1|	439585|	82738495|	2017|	3|	27579498|
|2|	146654|	651753738|	2017|	26|	25067451|
|3|	430504|	107277356|	2017|	6|	17879559|
|4|	294649|	3912359398|	2017|	265|	14763620|




### 2018년 1인당 매출 상위 기업( 39867 개 기업 중 )
|index|회사ID | 연매출액| 년도 | 연평균_직원수 | 1인당_매출액(천원) |
|--|-------|-------|-------|-------|-------|
|0|	567650|	113663897	|2018	|3|	37887965|
|1|	297874|	1115591447	|2018	|41| 27209547|
|2|	146654|	647641133	|2018	|25| 25905645|
|3|	40860| 170914439	|2018   |7|	24416348|
|4|	102971|	62943421	|2018	|3|	20981140|




### 2019년 1인당 매출 상위 기업( 852 개 기업 중 )
|index|회사ID | 연매출액| 년도 | 연평균_직원수 | 1인당_매출액(천원) |
|--|-------|-------|-------|-------|-------|
|0|	155754|	43590434|	2019|	11|	3962766|
|1|	419945|	10058868796|	2019|	2781|	3616997|
|2|	343863|	9892108|	2019|	4|	2473027|
|3|	212335|	7111215|	2019|	3|	2370405|
|4|	305146|	32791140|	2019|	16|	2049446|
